src/libraries/Microsoft.PowerFx.Core/Public/Types/TypeCheckers/AggregateTypeChecker.cs

```diff
@@ -0,0 +1,149 @@
+                {
+                    var extraFields = string.Join(", ", typeToCheck.FieldNames.Except(sourceType.FieldNames));
+                    _errorList.Add(new ExpressionError()
+                    {
+                        Kind = ErrorKind.Validation,
+                        Severity = ErrorSeverity.Critical,
+                        Message = $"Type mismatch between source and target record types. Given type has extra fields: {extraFields}."
+                    });
+                }
+                else
+                {
+                    var missingFields = string.Join(", ", sourceType.FieldNames.Except(typeToCheck.FieldNames));
+                    _errorList.Add(new ExpressionError()
+                    {
+                        Kind = ErrorKind.Validation,
+                        Severity = ErrorSeverity.Critical,
+                        Message = $"Type mismatch between source and target record types. Given type has missing fields: {missingFields}."
+                    });
+                }
+
+                _errorList.Add(new ExpressionError()
+                {
+                    Kind = ErrorKind.Validation,
+                    Severity = ErrorSeverity.Critical,
+                    Message = $"Type mismatch between source and target record types. SourceType field count: {sourceType.FieldNames.Count()}; TypeToCheck field count {typeToCheck.FieldNames.Count()}."
+                });
+                return false;
+            }
+
+            foreach (var sourceField in sourceType.FieldNames)
+            {
+                if (!sourceType.TryGetFieldType(sourceField, out var sourceFieldType))
+                {
+                    return false;
+                }
+
+                if (!typeToCheck.TryGetFieldType(sourceField, out var targetFieldType))
+                {
+                    _errorList.Add(new ExpressionError()
+                    {
+                        Kind = ErrorKind.Validation,
+                        Severity = ErrorSeverity.Critical,
+                        Message = $"Type mismatch between source and target record types. Field name: {sourceField} not found."
+                    });
+                    return false;
+                }
+
+                if (!IsFieldMatch(sourceFieldType, targetFieldType))
+                {
+                    _errorList.Add(new ExpressionError()
+                    {
+                        Kind = ErrorKind.Validation,
+                        Severity = ErrorSeverity.Critical,
+                        Message = $"Type mismatch between source and target record types. Field name: {sourceField} Expected {sourceFieldType._type.GetKindString()}; Found {targetFieldType._type.GetKindString()}."
+                    });
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        public override bool IsFieldMatch(FormulaType sourceField, FormulaType fieldToCheck)
```

Comment:
>IsFieldMatch [](http://example.com/codeflow?start=29&length=12)

This overlaps with FormulaTypeChecker.Run? 

What if IsMatch() took a FormulaTypeChecker, and then we just called on that instead of IsFieldMatch here?

Reply:
Updated